File: src/test/testData/dart/NullSafety.dart
// Dart-specific: ??, ??=, !, ?. — verify scoring matches plan decisions.

@complexity(1)
int? singleNullCoalesce(int? a, int? b) {
  return a ?? b;                         // +1 (?? scores like an OR sequence)
}

@complexity(1)
int? chainedNullCoalesce(int? a, int? b, int? c) {
  return a ?? b ?? c;                    // +1 (consecutive ?? = single sequence)
}

@complexity(2)
int? mixedOrAndCoalesce(int? a, int? b, bool? cond) {
  if (cond ?? false) {                   // +1 IF, +1 ?? sequence
    return a;
  }
  return b;
}

@complexity(0)
void compoundNullAssign(Map<String, int> m) {
  m["x"] ??= 0;                          // ??= is assignment, not a binary expression
}

@complexity(0)
void bangOperator(int? a) {
  print(a!);                             // ! null-assertion, not a decision
}

@complexity(0)
String? questionDotChain(String? s) {
  return s?.toLowerCase();               // ?. null-aware access, not a decision
}

@complexity(3)
int? guardWithCoalesce(int? a, int? b, int? c) {
  if (a == null && b == null) {          // +1 IF, +1 AND
    return c;
  }
  return a ?? b ?? c;                    // +1 ?? sequence
}

File: src/test/testData/kotlin/NullSafety.kt
// Mirrors dart/NullSafety.dart: elvis scores like Dart's ??, as an OR-style operator sequence.

@Complexity(1)
fun singleElvis(a: String?): String = a ?: "d"                  // +1

@Complexity(1)
fun chainedElvis(a: String?, b: String?): String = a ?: b ?: "d" // +1 (consecutive ?: = one sequence)

@Complexity(2)
fun elvisInsideIf(cond: Boolean?): String {
    if (cond ?: false) {                                        // +1 IF, +1 elvis
        return "y"
    }
    return "n"
}

@Complexity(2)
fun elvisThenOr(a: Boolean?, b: Boolean): Boolean = (a ?: false) || b // +1 elvis group, +1 OR

@Complexity(0)
fun safeCallChainIsFree(a: String?): Int? = a?.length?.dec()    // +0 (?. is not a decision)
